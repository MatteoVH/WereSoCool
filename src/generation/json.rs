use crate::{
    generation::nn_data_generation::{CSVOp, Normalizer},
    instrument::Basis,
    ui::{banner, printed},
    write::write_composition_to_json,
};
use num_rational::Rational64;
use serde::{Deserialize, Serialize};
use serde_json::to_string;
use socool_ast::{NormalForm, Normalize, OpOrNfTable, PointOp};

pub fn r_to_f64(r: Rational64) -> f64 {
    *r.numer() as f64 / *r.denom() as f64
}

#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
pub struct TimedOp {
    pub t: Rational64,
    pub event_type: EventType,
    pub voice: usize,
    pub event: usize,
    pub fm: Rational64,
    pub fa: Rational64,
    pub pm: Rational64,
    pub pa: Rational64,
    pub g: Rational64,
    pub l: Rational64,
}

fn normalize_value(value: Rational64, min: Rational64, max: Rational64) -> f64 {
    let result = (value - min) / (max - min);
    r_to_f64(result)
}

impl TimedOp {
    fn to_op_4d(&self, basis: &Basis) -> Op4D {
        Op4D {
            l: r_to_f64(self.l) * basis.l,
            t: r_to_f64(self.t) * basis.l,
            x: ((basis.p + r_to_f64(self.pa)) * r_to_f64(self.pm)),
            y: (basis.f * r_to_f64(self.fm)) + r_to_f64(self.fa),
            z: basis.g * r_to_f64(self.g),
            voice: self.voice,
            event: self.event,
            event_type: self.event_type.clone(),
        }
    }

    //    pub fn to_csv_op(&self) -> CSVOp {
    //        CSVOp {
    //            fm: r_to_f64(self.fm),
    //            fa: r_to_f64(self.fa),
    //            pm: r_to_f64(self.pm),
    //            pa: r_to_f64(self.pa),
    //            g: r_to_f64(self.g),
    //            l: r_to_f64(self.l),
    //            v: self.voice,
    //        }
    //    }

    pub fn to_csv_op(&self) -> CSVOp {
        let zero = Rational64::new(0, 1);
        let is_silent = self.fm == Rational64::new(0, 1) || self.g == Rational64::new(0, 1);
        let fm = if is_silent { zero } else { self.fm };
        let g = if is_silent { zero } else { self.g };

        CSVOp {
            fm: r_to_f64(fm),
            fa: r_to_f64(self.fa),
            pm: r_to_f64(self.pm),
            pa: r_to_f64(self.pa),
            g: r_to_f64(g),
            l: r_to_f64(self.l),
            v: self.voice as usize,
        }
    }

    //    pub fn to_normalized_csv_op(&self, normalizer: Normalizer) -> CSVOp {
    //        let zero = Rational64::new(0, 1);
    //        let is_silent = self.fm == Rational64::new(0, 1) || self.g == Rational64::new(0, 1);
    //        let fm = if is_silent { zero } else { self.fm };
    //        let g = if is_silent { zero } else { self.g };
    //
    //        CSVOp {
    //            fm: normalize_value(fm, normalizer.fm.0, normalizer.fm.1),
    //            fa: normalize_value(self.fa, normalizer.fa.0, normalizer.fa.1),
    //            pm: normalize_value(self.pm, normalizer.pm.0, normalizer.pm.1),
    //            pa: normalize_value(self.pa, normalizer.pa.0, normalizer.pa.1),
    //            g: normalize_value(g, normalizer.g.0, normalizer.g.1),
    //            l: normalize_value(self.l, normalizer.l.0, normalizer.l.1),
    //            v: normalize_value(
    //                Rational64::new(self.voice as i64, 1),
    //                normalizer.v.0,
    //                normalizer.v.1,
    //            ) as usize,
    //        }
    //    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Op4D {
    pub t: f64,
    pub event_type: EventType,
    pub voice: usize,
    pub event: usize,
    pub x: f64,
    pub y: f64,
    pub z: f64,
    pub l: f64,
}

#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
pub enum EventType {
    On,
    Off,
}

fn point_op_to_timed_op(
    point_op: &PointOp,
    time: &mut Rational64,
    voice: usize,
    event: usize,
) -> (TimedOp, TimedOp) {
    let on = TimedOp {
        fm: point_op.fm,
        fa: point_op.fa,
        pm: point_op.pm,
        pa: point_op.pa,
        g: point_op.g,
        l: point_op.l,
        t: time.clone(),
        event_type: EventType::On,
        voice,
        event,
    };

    *time += point_op.l;

    let off = TimedOp {
        t: time.clone(),
        event_type: EventType::Off,
        ..on
    };

    (on, off)
}

pub fn vec_timed_op_to_vec_op4d(timed_ops: Vec<TimedOp>, basis: &Basis) -> Vec<Op4D> {
    timed_ops.iter().map(|t_op| t_op.to_op_4d(&basis)).collect()
}

pub fn normalform_to_vec_timed_op_1d(normalform: &NormalForm, table: &OpOrNfTable) -> Vec<TimedOp> {
    let mut normal_form = NormalForm::init();

    println!("Generating Composition \n");
    normalform.apply_to_normal_form(&mut normal_form, table);

    let mut result: Vec<TimedOp> = normal_form
        .operations
        .iter()
        .enumerate()
        .flat_map(|(voice, vec_point_op)| {
            let mut time = Rational64::new(0, 1);
            let mut result = vec![];
            vec_point_op.iter().enumerate().for_each(|(event, p_op)| {
                let (on, _off) = point_op_to_timed_op(p_op, &mut time, voice, event);
                result.push(on);
                //                result.push(off);
            });
            result
        })
        .collect();

    result.sort_unstable_by_key(|a| a.t);

    result
}

pub fn normalform_to_vec_timed_op_2d(
    normalform: &NormalForm,
    table: &OpOrNfTable,
) -> Vec<Vec<TimedOp>> {
    let mut normal_form = NormalForm::init();

    println!("Generating Composition \n");
    normalform.apply_to_normal_form(&mut normal_form, table);

    let result: Vec<Vec<TimedOp>> = normal_form
        .operations
        .iter()
        .enumerate()
        .map(|(voice, vec_point_op)| {
            let mut time = Rational64::new(0, 1);
            let mut result = vec![];
            vec_point_op.iter().enumerate().for_each(|(event, p_op)| {
                let (on, _) = point_op_to_timed_op(p_op, &mut time, voice, event);
                result.push(on);
            });
            result
        })
        .collect();

    result
}

pub fn to_json(basis: &Basis, composition: &NormalForm, table: &OpOrNfTable, filename: String) {
    banner("JSONIFY-ing".to_string(), filename.clone());

    let vec_timed_op = normalform_to_vec_timed_op_1d(composition, table);
    let vec_op4d = vec_timed_op_to_vec_op4d(vec_timed_op, basis);

    let json = to_string(&vec_op4d).unwrap();

    write_composition_to_json(&json, &filename).expect("Writing to JSON failed");
    printed("JSON".to_string());
}
