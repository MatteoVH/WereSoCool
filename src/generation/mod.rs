pub mod nn_data_generator;
pub mod parsed_to_render;
mod test;
pub mod video_data_generation;
