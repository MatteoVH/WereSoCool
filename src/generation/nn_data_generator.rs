use num_rational::Rational64;
use socool_ast::ast::{OpOrNf::*, OpOrNfTable, Op::*, Op};
use socool_ast::operations::{NormalForm, Normalize as NormalizeOp, PointOp};
use insta::assert_debug_snapshot_matches;

pub fn r_to_f64(r: Rational64) -> f64 {
    *r.numer() as f64 / *r.denom() as f64
}

#[derive(Debug, Clone)]
pub struct DataOp {
    fm: f64,
    fa: f64,
    pm: f64,
    pa: f64,
    g: f64,
    l: f64,
}

#[derive(Debug, Clone)]
pub struct CSVOp {
    fm: f64,
    fa: f64,
    pm: f64,
    pa: f64,
    g: f64,
    t: f64,
}

impl DataOp {
    pub fn to_csv_op(&self, time: f64) -> CSVOp {
        CSVOp {
            fm: self.fm,
            fa: self.fa,
            pm: self.pm,
            pa: self.pa,
            g: self.g,
            t: time,
        }
    }
}

pub type NormalData = Vec<Vec<DataOp>>;
pub type CSVData = Vec<Vec<CSVOp>>;

pub fn point_op_to_data_op(
    point_op: &PointOp,
) -> DataOp {
    let result = DataOp {
        fm: r_to_f64(point_op.fm),
        fa: r_to_f64(point_op.fa),
        pm: r_to_f64(point_op.pm),
        pa: r_to_f64(point_op.pa),
        g: r_to_f64(point_op.g),
        l: r_to_f64(point_op.l),
    };

    result
}

pub fn composition_to_normal_data(composition: &NormalForm, table: &OpOrNfTable) -> NormalData {
    let mut normal_form = NormalForm::init();

    println!("Generating Composition \n");
    composition.apply_to_normal_form(&mut normal_form, table);

    let result: NormalData = normal_form
        .operations
        .iter()
        .map(|vec_point_op| {
            let mut result = vec![];
            for op in vec_point_op {
                let data_op = point_op_to_data_op(op);
                result.push(data_op);
            }
            result
        })
        .collect();

    result
}

pub fn normal_data_to_csv_data(data: NormalData) -> CSVData {
    let subdivision = 0.5;
    let result: CSVData = data
        .iter()
        .map(|vec_data_op| {
            let mut remainder = 0.0;
            let mut seq_time = 0.0;
            let mut result = vec![];
            for op in vec_data_op {
                let op_time = op.l;
                let n_steps_to_push = (op_time + remainder)/subdivision;
                remainder = n_steps_to_push.fract();
                for n in 0..n_steps_to_push.floor() as usize {
                    let new_op =  op.to_csv_op(seq_time);
                    result.push(new_op);
                    seq_time += subdivision;
                }
            }
            result
        })
        .collect();

    result
}

#[test]
fn normal_form_to_normal_data_test() {
    let mut normal_form = NormalForm::init();
    let pt = OpOrNfTable::new();

    Overlay {
        operations: vec![
            Op(Sequence {
                operations: vec![
                    Op(PanA {
                         a: Rational64::new(1, 2),
                    }),
                    Op(TransposeM {
                        m: Rational64::new(2, 1),
                    }),
                    Op(Gain {
                        m: Rational64::new(1, 2),
                    }),
                    Op(Length {
                        m: Rational64::new(2, 1),
                    }),
                ],
            }),
            Op(Sequence {
                operations: vec![Op(Length {
                    m: Rational64::new(5, 1),
                })],
            }),
        ],
    }
        .apply_to_normal_form(&mut normal_form, &pt);

    let normal_data = composition_to_normal_data(&normal_form, &pt);

    assert_debug_snapshot_matches!("normal_form_to_normal_data_test", normal_data);

    let normal_data_subdivided = normal_data_to_csv_data(normal_data);
    assert_debug_snapshot_matches!("normal_form_to_subdivided_test", normal_data_subdivided);
}
