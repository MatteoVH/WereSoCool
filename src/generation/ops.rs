use crate::{
    instrument::Basis,
    ui::{banner, printed},
    write::{write_composition_to_csv, write_composition_to_json, write_normalizer_to_json},
};
use num_rational::Rational64;
use serde::{Deserialize, Serialize};
use serde_json::to_string;
use socool_ast::{NormalForm, Normalize, OpOrNfTable, PointOp};
pub fn r_to_f64(r: Rational64) -> f64 {
    *r.numer() as f64 / *r.denom() as f64
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Op4D {
    pub t: f64,
    pub event_type: EventType,
    pub voice: usize,
    pub event: usize,
    pub x: f64,
    pub y: f64,
    pub z: f64,
    pub l: f64,
}

impl Op4D {
    pub fn normalize(&mut self, normalizer: &Normalizer) {
        self.x = 2.0 * normalize_value(self.x, normalizer.x.min, normalizer.x.max) - 1.0;
        self.y = normalize_value(self.y, normalizer.y.min, normalizer.y.max);
        self.z = normalize_value(self.z, normalizer.z.min, normalizer.z.max);
    }

    pub fn to_op_csv_1d(&self) -> OpCsv1d {
        OpCsv1d {
            time: self.t,
            length: self.l,
            frequency: self.y,
            pan: self.x,
            gain: self.z,
            voice: self.voice,
            event: self.event,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OpCsv1d {
    pub time: f64,
    pub length: f64,
    pub frequency: f64,
    pub pan: f64,
    pub gain: f64,
    pub voice: usize,
    pub event: usize,
}

impl OpCsv1d {
    pub fn to_op4d(&self) -> Op4D {
        Op4D {
            t: self.time,
            event_type: EventType::On,
            voice: self.voice,
            event: self.event,
            x: self.pan,
            y: self.frequency,
            z: self.gain,
            l: self.length,
        }
    }

    pub fn denormalize(&mut self, normalizer: &NormalizerJson) {
        let n = &normalizer.normalizer;
        self.pan = denormalize_value(self.pan, -1.0, 1.0, n.x.min, n.x.max);
        self.frequency = denormalize_value(self.frequency, 0.0, 1.0, n.y.min, n.y.max);
        self.gain = denormalize_value(self.gain, 0.0, 1.0, n.z.min, n.z.max);
    }
}

#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
pub struct TimedOp {
    pub t: Rational64,
    pub event_type: EventType,
    pub voice: usize,
    pub event: usize,
    pub fm: Rational64,
    pub fa: Rational64,
    pub pm: Rational64,
    pub pa: Rational64,
    pub g: Rational64,
    pub l: Rational64,
}

impl TimedOp {
    fn to_op_4d(&self, basis: &Basis) -> Op4D {
        let zero = Rational64::new(0, 1);
        let is_silent = (self.fm == zero && self.fa < Rational64::new(40, 1)) || self.g == zero;
        let y = if is_silent {
            0.0
        } else {
            (basis.f * r_to_f64(self.fm)) + r_to_f64(self.fa)
        };
        let z = if is_silent {
            0.0
        } else {
            basis.g * r_to_f64(self.g)
        };
        Op4D {
            l: r_to_f64(self.l) * basis.l,
            t: r_to_f64(self.t) * basis.l,
            x: ((basis.p + r_to_f64(self.pa)) * r_to_f64(self.pm)),
            y: y.log10(),
            z,
            voice: self.voice,
            event: self.event,
            event_type: self.event_type.clone(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Normalizer {
    pub x: MinMax,
    pub y: MinMax,
    pub z: MinMax,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MinMax {
    pub min: f64,
    pub max: f64,
}
