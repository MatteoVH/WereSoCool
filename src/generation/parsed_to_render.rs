extern crate num_rational;
extern crate pbr;
extern crate rayon;
extern crate serde;
extern crate serde_json;
extern crate socool_ast;
extern crate socool_parser;
use instrument::{
    oscillator::{Basis, Oscillator},
    stereo_waveform::{Normalize, StereoWaveform},
};
use num_rational::Rational64;
use pbr::ProgressBar;
use rayon::prelude::*;
use render::{Render, RenderPointOp};
use serde::{Deserialize, Serialize};
use serde_json::to_string;
use settings::default_settings;
use socool_ast::ast::{Op::*, OpOrNf::*, OpOrNfTable};
use socool_ast::operations::{NormalForm, Normalize as NormalizeOp, PointOp};
use std::sync::{Arc, Mutex};
use ui::{banner, printed};
use write::{write_composition_to_json, write_composition_to_wav};

pub fn r_to_f64(r: Rational64) -> f64 {
    *r.numer() as f64 / *r.denom() as f64
}

pub fn render(origin: &Basis, composition: &NormalForm, table: &OpOrNfTable) -> StereoWaveform {
    let mut normal_form = NormalForm::init();

    println!("\nGenerating Composition ");
    composition.apply_to_normal_form(&mut normal_form, table);
    //    println!("{:#?}", normal_form);

    let vec_wav = generate_waveforms(&origin, normal_form.operations, true);
    let mut result = sum_all_waveforms(vec_wav);
    result.normalize();

    result
}

pub fn render_mic(point_op: &PointOp, origin: Basis, osc: &mut Oscillator) -> StereoWaveform {
    let result = point_op.clone().render(&origin, osc, None);
    result
}

pub fn to_wav(composition: StereoWaveform, filename: String) {
    banner("Printing".to_string(), filename);
    write_composition_to_wav(composition);
    printed("WAV".to_string());
}

#[derive(Debug, Clone, PartialEq)]
pub enum EventType {
    On,
    Off
}

#[derive(Debug, Clone, PartialEq)]
pub struct Op4D {
    x: f64,
    y: f64,
    z: f64,
    t: Rational64,
    event_type: EventType,
    voice: usize,
    event: usize,
}

fn point_op_to_4d(point_op: &PointOp, basis: &Basis, time: &mut Rational64, voice: usize, event: usize) -> Op4D {
    let result = Op4D {
        x: basis.p * r_to_f64(point_op.pa) + r_to_f64(point_op.pa),
        y: basis.f * r_to_f64(point_op.fm) + r_to_f64(point_op.fa),
        z: basis.g * r_to_f64(point_op.g),
        t: time.clone(),
        event_type: EventType::On,
        voice,
        event,
    };

    *time += point_op.l;

    result
}

fn composition_to_vec_op4d(
    basis: &Basis,
    composition: &NormalForm,
    table: &OpOrNfTable,
) -> Vec<Vec<Op4D>> {
    let mut normal_form = NormalForm::init();

    println!("Generating Composition \n");
    composition.apply_to_normal_form(&mut normal_form, table);

    normal_form
        .operations
        .iter()
        .enumerate()
        .map(|(voice, vec_point_op)| {
            let mut time = Rational64::new(0, 1);
            vec_point_op
                .iter()
                .enumerate()
                .map(|(event, p_op)| point_op_to_4d(p_op, basis, &mut time, voice, event))
                .collect()
        })
        .collect()
}

pub fn to_json(basis: &Basis, composition: &NormalForm, table: &OpOrNfTable, filename: String) {
    banner("JSONIFY-ing".to_string(), filename.clone());

    let vec_op4d = composition_to_vec_op4d(basis, composition, table);

//    let json = to_string(&vec_op4d).unwrap();
    //    json
//    write_composition_to_json(&json, &filename).expect("Writing to JSON failed");
//    printed("JSON".to_string());
}

fn create_pb_instance(n: usize) -> Arc<Mutex<ProgressBar<std::io::Stdout>>> {
    let mut pb = ProgressBar::new(n as u64);
    pb.format("╢w♬░╟");
    pb.message("Rendering:  ");
    Arc::new(Mutex::new(pb))
}

pub fn generate_waveforms(
    origin: &Basis,
    mut vec_sequences: Vec<Vec<PointOp>>,
    show: bool,
) -> Vec<StereoWaveform> {
    if show {
        println!("Rendering {:?} waveforms", vec_sequences.len());
    }
    let pb = create_pb_instance(vec_sequences.len());

    let vec_wav = vec_sequences
        .par_iter_mut()
        .map(|ref mut vec_point_op: &mut Vec<PointOp>| {
            pb.lock().unwrap().add(1 as u64);
            let mut osc = Oscillator::init(&default_settings());
            vec_point_op.render(&origin, &mut osc)
        })
        .collect();

    pb.lock().unwrap().finish_print(&"".to_string());

    vec_wav
}

pub fn sum_all_waveforms(mut vec_wav: Vec<StereoWaveform>) -> StereoWaveform {
    let mut result = StereoWaveform::new(0);

    sort_vecs(&mut vec_wav);

    let max_len = vec_wav[0].l_buffer.len();

    result.l_buffer.resize(max_len, 0.0);
    result.r_buffer.resize(max_len, 0.0);

    for wav in vec_wav {
        sum_vec(&mut result.l_buffer, &wav.l_buffer[..]);
        sum_vec(&mut result.r_buffer, &wav.r_buffer[..])
    }

    result
}

fn sort_vecs(vec_wav: &mut Vec<StereoWaveform>) {
    vec_wav.sort_unstable_by(|a, b| b.l_buffer.len().cmp(&a.l_buffer.len()));
}

fn sum_vec(a: &mut Vec<f64>, b: &[f64]) {
    for (ai, bi) in a.iter_mut().zip(b) {
        *ai += *bi;
    }
}

#[cfg(test)]
pub mod tests {
    use super::*;
    #[test]
    fn render_equal() {
        let mut a = vec![1.0, 2.0, 3.0];
        let b = vec![1.0, 2.0, 3.0];
        sum_vec(&mut a, &b[..]);
        let expected = [2.0, 4.0, 6.0];
        assert_eq!(a, expected);
    }

    #[test]
    fn render_left() {
        let mut a = vec![1.0, 2.0, 3.0, 2.0];
        let b = vec![1.0, 2.0, 3.0];
        sum_vec(&mut a, &b[..]);
        let expected = [2.0, 4.0, 6.0, 2.0];
        assert_eq!(a, expected);
    }
    #[test]
    fn to_vec_op4d_test() {
        let mut normal_form = NormalForm::init();
        let pt = OpOrNfTable::new();
        let basis = Basis {
            f: 100.0,
            g: 1.0,
            p: 0.0,
            l: 1.0,
            a: 44100.0,
            d: 44100.0,
        };

        Sequence {
            operations: vec![
                Op(PanA {
                    a: Rational64::new(1, 2),
                }),
                Op(TransposeM {
                    m: Rational64::new(2, 1),
                }),
                Op(Gain {
                    m: Rational64::new(1, 2),
                }),
                Op(Length {
                    m: Rational64::new(2, 1),
                }),
            ],
        }
            .apply_to_normal_form(&mut normal_form, &pt);

        let result = composition_to_vec_op4d(&basis, &normal_form, &pt);
        assert_eq!(
            result,
            vec![vec![
                Op4D {
                    x: 0.5,
                    y: 100.0,
                    z: 1.0,
                    t: Rational64::new(0, 1),
                    event_type: EventType::On,
                    voice: 0,
                    event: 0,
                },
                Op4D {
                    x: 0.5,
                    y: 100.0,
                    z: 1.0,
                    t: Rational64::new(1, 1),
                    event_type: EventType::Off,
                    voice: 0,
                    event: 0,
                },

                Op4D {
                    x: 0.0,
                    y: 200.0,
                    z: 1.0,
                    t: Rational64::new(1, 1),
                    event_type: EventType::On,
                    voice: 0,
                    event: 1,
                },
                Op4D {
                    x: 0.0,
                    y: 200.0,
                    z: 1.0,
                    t: Rational64::new(2, 1),
                    event_type: EventType::Off,
                    voice: 0,
                    event: 1,
                },

                Op4D {
                    x: 0.0,
                    y: 100.0,
                    z: 0.5,
                    t: Rational64::new(2, 1),
                    event_type: EventType::On,
                    voice: 0,
                    event: 2,
                },
                Op4D {
                    x: 0.0,
                    y: 100.0,
                    z: 0.5,
                    t: Rational64::new(3, 1),
                    event_type: EventType::Off,
                    voice: 0,
                    event: 2,
                },

                Op4D {
                    x: 0.0,
                    y: 100.0,
                    z: 1.0,
                    t: Rational64::new(3, 1),
                    event_type: EventType::On,
                    voice: 0,
                    event: 3,
                },
                Op4D {
                    x: 0.0,
                    y: 100.0,
                    z: 1.0,
                    t: Rational64::new(5, 1),
                    event_type: EventType::Off,
                    voice: 0,
                    event: 3,
                },
            ]]
        )
    }
}

