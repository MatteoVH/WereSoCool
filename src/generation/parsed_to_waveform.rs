extern crate num_rational;
extern crate socool_parser;
use event::{Event, Render};
use instrument::{oscillator::Oscillator, stereo_waveform::StereoWaveform};
use num_rational::Rational;
use operations::Apply;
use settings::get_default_app_settings;
use socool_parser::{ast::Op, parser::Init};

fn oscillator() -> Oscillator {
    Oscillator::init(&get_default_app_settings())
}

pub fn r_to_f32(r: Rational) -> f32 {
//    let x = 100000000000000000000000000000000000000;
//    if r.numer() > &x || r.denom() > &x {
//        let mut fract = r.fract();
//        for _ in 0..32 {
//            if fract.is_integer() {
//                break;
//            }
//            fract = fract * 10
//        }
//
//        return r.to_integer() as f32 + fract.to_integer() as f32;
//    }

    simple_r_to_f32(r)
}

fn simple_r_to_f32(r: Rational) -> f32 {
    *r.numer() as f32 / *r.denom() as f32
}

pub fn event_from_init(init: Init) -> Event {
    Event::init(
        r_to_f32(init.f),
        r_to_f32(init.g),
        r_to_f32(init.p),
        r_to_f32(init.l),
    )
}

pub fn generate_events(init: Init, main: Op) -> Vec<Event> {
    main.apply(vec![event_from_init(init)])
}

pub fn generate_composition(init: Init, main: Op) -> StereoWaveform {
    generate_events(init, main).render(&mut oscillator())
}
