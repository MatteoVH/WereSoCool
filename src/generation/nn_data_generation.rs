use crate::generation::{normalform_to_vec_timed_op_1d, TimedOp};
use insta::assert_debug_snapshot_matches;
use num_rational::Rational64;
use socool_ast::OpOrNf;
use socool_parser::parser::*;

#[derive(Debug, Clone)]
pub struct CSVOp {
    pub fm: f64,
    pub fa: f64,
    pub pm: f64,
    pub pa: f64,
    pub g: f64,
    pub l: f64,
    pub v: usize,
}

#[derive(Debug, Clone)]
pub struct Normalizer {
    pub fm: (Rational64, Rational64),
    pub fa: (Rational64, Rational64),
    pub pm: (Rational64, Rational64),
    pub pa: (Rational64, Rational64),
    pub g: (Rational64, Rational64),
    pub l: (Rational64, Rational64),
    pub v: (Rational64, Rational64),
}

pub type CSVData = Vec<CSVOp>;

pub fn vec_timed_op_to_vec_csv_data(timed_ops: Vec<TimedOp>) -> CSVData {
    timed_ops.iter().map(|t_op| t_op.to_csv_op()).collect()
}

pub fn vec_timed_op_1d_to_csv_data_1d(timed_ops: Vec<TimedOp>) -> Vec<CSVOp> {
    timed_ops.iter().map(|t_op| t_op.to_csv_op()).collect()
}

pub fn vec_timed_op_2d_to_csv_data_2d(timed_ops: Vec<Vec<TimedOp>>) -> Vec<Vec<CSVOp>> {
    let result: Vec<Vec<CSVOp>> = timed_ops
        .iter()
        .map(|vec_t_op| vec_t_op.iter().map(|t_op| t_op.to_csv_op()).collect())
        .collect();
    result
}

//pub fn csv_data_to_normalized_csv_data(data: CSVData, ) {
//    let max__state = CSVOp {
//        fm: 70.0,
//        fa: 33.0,
//        pm: 1.0,
//        pa: 1.7857142857142858,
//        g: 1.0,
//        l: 576.0,
//        v: 191
//    }
//
//}

fn get_max_min_csv_data(csv_data: CSVData) -> (CSVOp, CSVOp) {
    let mut max_state = CSVOp {
        fm: 0.0,
        fa: 0.0,
        pm: 0.0,
        pa: 0.0,
        g: 0.0,
        l: 0.0,
        v: 0,
    };
    let mut min_state = CSVOp {
        fm: 0.0,
        fa: 0.0,
        pm: 0.0,
        pa: 0.0,
        g: 0.0,
        l: 0.0,
        v: 0,
    };

    csv_data.iter().for_each(|csv_op| {
        max_state = CSVOp {
            fm: max_state.fm.max(csv_op.fm),
            fa: max_state.fa.max(csv_op.fa),
            pm: max_state.pm.max(csv_op.pm),
            pa: max_state.pa.max(csv_op.pa),
            g: max_state.g.max(csv_op.g),
            l: max_state.l.max(csv_op.l),
            v: max_state.v.max(csv_op.v),
        };
        min_state = CSVOp {
            fm: min_state.fm.min(csv_op.fm),
            fa: min_state.fa.min(csv_op.fa),
            pm: min_state.pm.min(csv_op.pm),
            pa: min_state.pa.min(csv_op.pa),
            g: min_state.g.min(csv_op.g),
            l: min_state.l.min(csv_op.l),
            v: min_state.v.min(csv_op.v),
        };
    });

    (max_state, min_state)
}

pub fn get_min_max_for_path(filename: String) -> (CSVOp, CSVOp, usize) {
    let parsed = parse_file(&filename.to_string(), None);
    let parsed_main = parsed.table.get("main").unwrap();

    let nf = match parsed_main {
        OpOrNf::Nf(nf) => nf,
        OpOrNf::Op(_) => panic!("main is Not in Normal Form for some terrible reason."),
    };

    let timed_ops = normalform_to_vec_timed_op_1d(nf, &parsed.table);
    let n_voices = timed_ops.len();
    let csv_data = vec_timed_op_1d_to_csv_data_1d(timed_ops);
    dbg!(csv_data.clone());
    let (max, min) = get_max_min_csv_data(csv_data);

    (max, min, n_voices)
}

#[cfg(test)]
mod nn_data_test {
    use super::*;
    use socool_ast::{NormalForm, Normalize, Op::*, OpOrNf::*, OpOrNfTable, PointOp};
    use walkdir::WalkDir;

    //    #[test]
    //    fn normal_form_to_normal_data_test() {
    //        let mut normal_form = NormalForm::init();
    //        let pt = OpOrNfTable::new();
    //
    //        Overlay {
    //            operations: vec![
    //                Op(Sequence {
    //                    operations: vec![
    //                        Op(PanA {
    //                            a: Rational64::new(-1, 2),
    //                        }),
    //                        Op(TransposeM {
    //                            m: Rational64::new(2, 1),
    //                        }),
    //                        Op(Gain {
    //                            m: Rational64::new(1, 2),
    //                        }),
    //                        Op(Length {
    //                            m: Rational64::new(2, 1),
    //                        }),
    //                    ],
    //                }),
    //                Op(Sequence {
    //                    operations: vec![Op(Length {
    //                        m: Rational64::new(5, 1),
    //                    })],
    //                }),
    //            ],
    //        }
    //        .apply_to_normal_form(&mut normal_form, &pt);
    //
    //        let normal_data = normalform_to_vec_timed_op_1d(&normal_form, &pt);
    //
    //        assert_debug_snapshot_matches!("normal_form_to_normal_data_test", normal_data);
    //
    //        let csv_data = vec_timed_op_to_vec_csv_data(normal_data);
    //        assert_debug_snapshot_matches!("normal_form_to_subdivided_test", csv_data);
    //
    //        let (max, min) = get_max_min_csv_data(csv_data);
    //        assert_debug_snapshot_matches!("max_csv_data", max);
    //        assert_debug_snapshot_matches!("min_csv_data", min);
    //    }

    #[test]
    fn test_csv_of_file() {
        let mut max_state = CSVOp {
            fm: 0.0,
            fa: 0.0,
            pm: 0.0,
            pa: 0.0,
            g: 0.0,
            l: 0.0,
            v: 0,
        };
        let mut min_state = CSVOp {
            fm: 0.0,
            fa: 0.0,
            pm: 0.0,
            pa: 0.0,
            g: 0.0,
            l: 0.0,
            v: 0,
        };

        for entry in WalkDir::new("./songs/test_data/")
            .follow_links(true)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let f_name = entry.path().to_string_lossy();

            if f_name.ends_with(".socool") && !f_name.contains("import") {
                let (song_max, song_min, n_events) = get_min_max_for_path(f_name.to_string());
                println!("{}", f_name);
                max_state = CSVOp {
                    fm: max_state.fm.max(song_max.fm),
                    fa: max_state.fa.max(song_max.fa),
                    pm: max_state.pm.max(song_max.pm),
                    pa: max_state.pa.max(song_max.pa),
                    g: max_state.g.max(song_max.g),
                    l: max_state.l.max(song_max.l),
                    v: max_state.v.max(song_max.v),
                };
                min_state = CSVOp {
                    fm: min_state.fm.min(song_min.fm),
                    fa: min_state.fa.min(song_min.fa),
                    pm: min_state.pm.min(song_min.pm),
                    pa: min_state.pa.min(song_min.pa),
                    g: min_state.g.min(song_min.g),
                    l: min_state.l.min(song_min.l),
                    v: min_state.v.min(song_min.v),
                };
            }
        }

        assert_debug_snapshot_matches!("test_csv_of_file_max_for_path", max_state);
        assert_debug_snapshot_matches!("test_csv_of_file_min_for_path", min_state);
    }

}
