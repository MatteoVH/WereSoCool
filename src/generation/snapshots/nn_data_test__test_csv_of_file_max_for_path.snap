---
created: "2019-05-23T04:12:40.952202Z"
creator: insta@0.8.1
source: src/generation/nn_data_generation.rs
expression: max_state
---
CSVOp {
    fm: 4.0,
    fa: 0.0,
    pm: 1.0,
    pa: 1.0,
    g: 1.0,
    l: 8.0,
    v: 1
}
